File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.cpp
//VERY IMPO PROBLEM = ASKED IN META
//SOLUTION BY SORTING WITH THE CUSTOM COMPARATOR
// class Solution {
// public:
//     vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
//         sort(points.begin(),points.end(),[&](vector<int> a,vector<int> b){
//             return squaredDist(a)<squaredDist(b);
//         });
        
//         return vector<vector<int>>(points.begin(),points.begin()+k);
//     }
    
// private:
//     int squaredDist(vector<int>& point)
//     {
//         return point[0]*point[0]+point[1]*point[1];
//     }
// };


//USING MAX HEAP IE PRIORITY QUEUE = Their solution 
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        
        priority_queue<pair<int,int>> maxheap;
        
        for(int i=0;i<points.size();i++)
        {
            pair<int,int> entry = {squaredDist(points[i]),i};
            
            if(k>maxheap.size())
            {
                maxheap.push(entry);
            }
            else if(entry.first<maxheap.top().first) //IMPO TO ADD THIS CONDITION SINCE IF WE HAVE TO CHECK IF THE ELEMENT TO BE PUSHED IS CLOSER THAN THE FARTHEST ELEMENT IN THE HEAP
            {
                maxheap.pop();
                maxheap.push(entry);
            }
        }
        
        vector<vector<int>> ans;
        while(!maxheap.empty())
        {
            ans.push_back(points[maxheap.top().second]);
            maxheap.pop();
        }
        return ans;
    }

private:
    int squaredDist(vector<int>& point)
    {
        return point[0]*point[0]+point[1]*point[1];
    }
};

File: 973-k-closest-points-to-origin/973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        
        PriorityQueue<Pair<Integer,Integer>> maxh =  new PriorityQueue<>((a,b)->b.getKey()-a.getKey());
        
        for(int i=0;i<points.length;i++)
        {
            Pair<Integer,Integer> entry = new Pair<>(squaredDist(points[i]),i);
            maxh.add(entry);
            if(maxh.size()>k)
                maxh.poll();  
        }
        
        int result[][] = new int[k][2];
        int i=0;
        while(maxh.size()>0)
        {
            int l = maxh.poll().getValue();
            result[i][0] = points[l][0];
            result[i][1] = points[l][1];
            i++;
        }
        return result;
    }
    
    private int squaredDist(int[] point)
    {
        return point[0]*point[0]+point[1]*point[1];
    }
}